libp2p/transports/tcptransport.nim

```diff
@@ -144,34 +137,65 @@ method close*(t: TcpTransport) {.async, gcsafe.} =
 
     trace "transport stopped"
     inc getTcpTransportTracker().closed
-  except CancelledError as exc:
-    raise exc
   except CatchableError as exc:
-    trace "error shutting down tcp transport", exc = exc.msg
-
-method listen*(t: TcpTransport,
-               ma: MultiAddress,
-               handler: ConnHandler):
-               Future[Future[void]] {.async, gcsafe.} =
-  discard await procCall Transport(t).listen(ma, handler) # call base
+    let useExc {.used.} = exc
+    trace "error shutting down tcp transport", errMsg = exc.msg
+  finally:
+    t.running = false
 
-  ## listen on the transport
-  t.server = createStreamServer(t.ma, connCb, t.flags, t)
-  t.server.start()
-
-  # always get the resolved address in case we're bound to 0.0.0.0:0
-  t.ma = MultiAddress.init(t.server.sock.getLocalAddress()).tryGet()
-  result = t.server.join()
-  trace "started node on", address = t.ma
+method accept*(t: TcpTransport): Future[Connection] {.async, gcsafe.} =
+  try:
+    let transp = await t.server.accept()
+    try:
+      # we don't need result connection in this
+      # case as it's added inside connHandler
+      return t.connHandler(transp, false)
+    except CancelledError as exc:
+      raise exc
+    except CatchableError as exc:
+      let useExc {.used.} = exc
+      debug "listen: connection setup failed", errMsg = exc.msg
+      if not transp.closed:
+        await transp.closeWait()
+  except TransportTooManyError as exc:
+    let useExc {.used.} = exc
+    warn "listen: could not accept new client, too many files opened"
```

Comment:
return `nil` here?

Reply:
Can be, to make it more explicit, but this will fall through and return `nil` either way for all code paths except the`try` block itself.

Reply:
it doesn't look intentional this way, ie more like a forgotten branch - returning nil vs exceptions remains confusing